### PR TITLE
Add support for offline and local builds

### PR DESCRIPTION
**Pull request by edvbld**
Hi all,

this patch adds much improved support for both _offline_ and _local_ builds. We are currently downloading both JDK and Gradle distributions for the most common platforms (GNU/Linux x64, macOS x64 and Windows x64), but this scheme does not scale to multiple other operating systems and/or CPU architectures. I have therefore extended the `images` plugin to support a `local` image which uses the JDK that is currently executing Gradle for producing the jlinked image.

I also added support for fully offline builds for those want to supply both a JDK and Gradle themselves and ensure that the build does not download anything.

Thanks,
Erik

## Testing
- Tested multiple variants of builds, including local, offline and default builds

## Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

## Approvers
 * Robin Westberg ([rwestberg](@rwestberg) - **Reviewer**) **Note!** Review applies to 66a6b56bbab2b8bd7d79f4c31e39c35a94850824

### CODE REVIEW

**Comment by edvbld** [comment]
@nick-arm could you please try this PR out and see if it works for you?

**Comment by edvbld** [comment]
@JornVernee could you give this a spin on Windows and see if it works?

File: README.md [hunk 1]
```diff
@@ -55,6 +55,42 @@ To build the project on Windows, run the following command from the source tree
...
+```bash
+$ sh gradlew
+```
+
+The extracted jlinked image will end up in the `build` directory in the source
+tree root.
+
+### Offline builds
+
+If you don't want the build to automatically download any depenendcies, then
```
Comment (rwestberg) [inline README.md]:
```suggestion
If you don't want the build to automatically download any dependencies, then
```

File: README.md [hunk 2]
```diff
@@ -55,6 +55,42 @@ To build the project on Windows, run the following command from the source tree
...
+you must ensure that you have installed the following software locally:
+
+- JDK 12 or later
+- Gradle 5.6.2 or later
+
+To create a build then run the command
```
Comment (rwestberg) [inline README.md]:
```suggestion
To create a build then run the command:
```

**Comment by JornVernee** [comment]
Works for me with gradle 5.6.4 and JDK 12b33

File: buildSrc/images/src/main/java/org/openjdk/skara/gradle/images/ImagesPlugin.java [hunk 3]
```diff
@@ -29,9 +29,49 @@
...
+        if (p.startsWith("sunos")) {
+            return "solaris";
+        }
+
+        throw new RuntimeException("Unknown operating system: " + System.getProperty("os.name"));
+    }
+
+    private static String getCPU() {
+        var p = System.getProperty("os.arch").toLowerCase();
+        if (p.startsWith("amd64")) {
```
[the remark this answers is not in the document]
Comment (edvbld) [reply buildSrc/images/src/main/java/org/openjdk/skara/gradle/images/ImagesPlugin.java]: Thanks, fixed! I also added a check for `"x64"`.

File: build.gradle [hunk 4]
```diff
@@ -104,19 +104,79 @@ reproduce {
...
+        return 'windows'
+    }
+    if (os.startsWith('sunos')) {
+        return 'solaris'
+    }
+    throw new GradleException("Unexpected operating system: " + os)
+}
+
+def getCPU() {
+    def cpu = System.getProperty('os.arch').toLowerCase()
```
[the remark this answers is not in the document]
Comment (edvbld) [reply build.gradle]: Fixed, also added a check for `"x64"`.

File: build.gradle [hunk 5]
```diff
@@ -104,19 +104,79 @@ reproduce {
...
+    if (cpu.startsWith('sparc')) {
+        return 'sparc'
+    }
+    if (cpu.startsWith('ppc')) {
+        return 'ppc'
+    }
+    if (cpu.startsWith('arm')) {
+        return 'arm'
+    }
+    throw new GradleException("Unexpected operating system: " + cpu)
```
Comment (nick-arm) [inline build.gradle]: Should be "Unexpected CPU"?

File: build.gradle [hunk 6]
```diff
@@ -104,19 +104,79 @@ reproduce {
...
+    }
+    if (cpu.startsWith('sparc')) {
+        return 'sparc'
+    }
+    if (cpu.startsWith('ppc')) {
+        return 'ppc'
+    }
+    if (cpu.startsWith('arm')) {
+        return 'arm'
+    }
```
Comment (nick-arm) [inline build.gradle]: 64-bit and 32-bit Arm have different `os.arch` values. So we need to add:

```groovy
if (cpu.startsWith('aarch64')) {
    return 'aarch64'
}
```

File: buildSrc/images/src/main/java/org/openjdk/skara/gradle/images/ImagesPlugin.java [hunk 7]
```diff
@@ -29,9 +29,49 @@
...
+        }
+        if (p.startsWith("sparc")) {
+            return "sparc";
+        }
+        if (p.startsWith("ppc")) {
+            return "ppc";
+        }
+        if (p.startsWith("arm")) {
+            return "arm";
+        }
```
Comment (nick-arm) [inline buildSrc/images/src/main/java/org/openjdk/skara/gradle/images/ImagesPlugin.java]: Also need to check for `aarch64` here:

```java
if (p.startsWith("aarch64")) {
    return "aarch64";
}
```

File: build.gradle (hunk 6 above)
Comment (edvbld) [reply build.gradle]: Ah, right, thanks. Fixed!

File: buildSrc/images/src/main/java/org/openjdk/skara/gradle/images/ImagesPlugin.java (hunk 7 above)
Comment (edvbld) [reply buildSrc/images/src/main/java/org/openjdk/skara/gradle/images/ImagesPlugin.java]: Thanks, fixed this one as well!

File: build.gradle (hunk 5 above)
Comment (edvbld) [reply build.gradle]: Yep, fixed, thanks!

**Pull request closed without merging**